src/nginx.c

```diff
@@ -73,108 +85,261 @@ static size_t nginx_curl_callback(void *buf, size_t size, size_t nmemb,
   }
 
-  if ((verify_host == NULL) || IS_TRUE(verify_host)) {
-    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 2L);
+  if ((parsed_values_for_nginx_instance[4][0] == '\0') ||
+      IS_TRUE(parsed_values_for_nginx_instance[4])) {
+    curl_easy_setopt(ngx_inst[instance_num_counter].curl,
+                     CURLOPT_SSL_VERIFYHOST, 2L);
   } else {
-    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
+    curl_easy_setopt(ngx_inst[instance_num_counter].curl,
+                     CURLOPT_SSL_VERIFYHOST, 0L);
   }
 
-  if (cacert != NULL) {
-    curl_easy_setopt(curl, CURLOPT_CAINFO, cacert);
+  if (parsed_values_for_nginx_instance[5][0] != '\0') {
+    curl_easy_setopt(ngx_inst[instance_num_counter].curl, CURLOPT_CAINFO,
+                     parsed_values_for_nginx_instance[5]);
   }
 
 #ifdef HAVE_CURLOPT_TIMEOUT_MS
-  if (timeout != NULL) {
-    curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS, atol(timeout));
+  if (parsed_values_for_nginx_instance[6][0] != '\0') {
+    curl_easy_setopt(ngx_inst[instance_num_counter].curl, CURLOPT_TIMEOUT_MS,
+                     atol(parsed_values_for_nginx_instance[6]));
   } else {
-    curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS,
+    curl_easy_setopt(ngx_inst[instance_num_counter].curl, CURLOPT_TIMEOUT_MS,
                      (long)CDTIME_T_TO_MS(plugin_get_interval()));
   }
 #endif
 
 #ifdef HAVE_CURLOPT_UNIX_SOCKET_PATH
-  if (sock != NULL) {
-    curl_easy_setopt(curl, CURLOPT_UNIX_SOCKET_PATH, sock);
+  if (parsed_values_for_nginx_instance[7][0] != '\0') {
+    curl_easy_setopt(ngx_inst[instance_num_counter].curl,
+                     CURLOPT_UNIX_SOCKET_PATH,
+                     parsed_values_for_nginx_instance[7]);
   }
 #endif
 
+  curl_easy_setopt(ngx_inst[instance_num_counter].curl, CURLOPT_URL,
+                   parsed_values_for_nginx_instance[0]);
+  return 0;
+}
+/******************/
+
+/* Initialize CURL structure for the conf and set up default options */
+static int init_curl_structure() {
+
+  if (ngx_inst[instance_num_counter].curl != NULL)
+    curl_easy_cleanup(ngx_inst[instance_num_counter].curl);
+
+  if ((ngx_inst[instance_num_counter].curl = curl_easy_init()) == NULL) {
+    ERROR("nginx plugin: curl_easy_init failed.");
+    return -1;
+  }
+
+  return 0;
+}
+/******************/
+
+/******* Nginx plugin config parsing. Inspired by ceph plugin config parser
+ * implementation ( see ceph.c ) *******/
+static int nginx_handle_str(struct oconfig_item_s *item, char *dest,
+                            int dest_len) {
+  const char *val;
+  if (item->values_num != 1) {
+    return -ENOTSUP;
+  }
+  if (item->values[0].type != OCONFIG_TYPE_STRING) {
+    return -ENOTSUP;
+  }
+  val = item->values[0].value.string;
+  if (snprintf(dest, dest_len, "%s", val) > (dest_len - 1)) {
+    ERROR("nginx plugin: configuration parameter '%s' is too long.\n",
+          item->key);
+    return -ENAMETOOLONG;
+  }
   return 0;
-} /* void init */
+}
```

Comment:
Earlier code did not use strings with fixed lengths, but took the strings from config as-is in `config_set()`.  I think the new code should do the same.